File: phases/00-sampling-fundamentals/01-intro-to-sampling/code/main.py
# ============================================================
# Lesson 0.1: Verifying Unbiasedness via Simulation
# Python — From Scratch (No specialized libraries)
# ============================================================

import numpy as np
import pandas as pd

np.random.seed(42)

# --- Generate a small population ---
N = 10000
population_income = np.random.lognormal(mean=7.5, sigma=0.9, size=N)

# True population parameter
theta = np.mean(population_income)
print(f"True population mean (theta): {theta:.4f}")

# --- Repeated sampling experiment ---
n = 200
B = 5000

estimates = np.zeros(B)

for b in range(B):
    idx = np.random.choice(N, size=n, replace=False)
    sample_data = population_income[idx]
    estimates[b] = np.mean(sample_data)

# --- Verify unbiasedness ---
expected_value = np.mean(estimates)
bias = expected_value - theta
variance_est = np.var(estimates, ddof=1)
mse = variance_est + bias**2

print(f"\n--- Simulation Results (B = {B}) ---")
print(f"E(theta_hat)   : {expected_value:.4f}")
print(f"theta (true)   : {theta:.4f}")
print(f"Bias           : {bias:.4f}")
print(f"Variance       : {variance_est:.4f}")
print(f"MSE            : {mse:.4f}")

# --- Theoretical variance under SRSWOR ---
S2 = np.var(population_income, ddof=1)
theoretical_var = (1 - n / N) * S2 / n
print(f"\nTheoretical Var: {theoretical_var:.4f}")
print(f"Simulated Var  : {variance_est:.4f}")
print(f"Ratio          : {variance_est / theoretical_var:.4f} (should be ~1.0)")

# Assertion
assert abs(bias) < 0.5 * np.sqrt(theoretical_var), "Bias too large!"
print("\n[PASS] Unbiasedness verified.")

# ============================================================
# Production: samplics / weighted estimation comparison
# ============================================================

population_df = pd.DataFrame({
    'id': range(1, N + 1),
    'income': population_income
})

sample_df = population_df.sample(n=n, random_state=42).copy()
sample_df['weight'] = N / n

manual_mean = sample_df['income'].mean()
weighted_mean = np.average(sample_df['income'], weights=sample_df['weight'])

print(f"\nManual mean   : {manual_mean:.4f}")
print(f"Weighted mean : {weighted_mean:.4f}")

assert abs(manual_mean - weighted_mean) < 1e-10
print("[PASS] Manual and weighted estimates match.")

File: phases/00-sampling-fundamentals/01-intro-to-sampling/code/main.R
# ============================================================
# Lesson 0.1: Verifying Unbiasedness via Simulation
# R — From Scratch (No survey packages)
# ============================================================

set.seed(42)

# --- Generate a small population ---
N <- 10000
population <- data.frame(
  id     = 1:N,
  income = rlnorm(N, meanlog = 7.5, sdlog = 0.9)
)

# True population parameter (theta)
theta <- mean(population$income)
cat(sprintf("True population mean (theta): %.4f\n", theta))

# --- Repeated sampling experiment ---
n <- 200           # Sample size
B <- 5000          # Number of replications

estimates <- numeric(B)

for (b in 1:B) {
  idx <- sample(1:N, n, replace = FALSE)
  sample_data <- population$income[idx]
  estimates[b] <- mean(sample_data)
}

# --- Verify unbiasedness ---
expected_value <- mean(estimates)
bias <- expected_value - theta
variance_est <- var(estimates)
mse <- variance_est + bias^2

cat(sprintf("\n--- Simulation Results (B = %d) ---\n", B))
cat(sprintf("E(theta_hat)   : %.4f\n", expected_value))
cat(sprintf("theta (true)   : %.4f\n", theta))
cat(sprintf("Bias           : %.4f\n", bias))
cat(sprintf("Variance       : %.4f\n", variance_est))
cat(sprintf("MSE            : %.4f\n", mse))

# --- Theoretical variance under SRSWOR ---
S2 <- var(population$income)
theoretical_var <- (1 - n/N) * S2 / n
cat(sprintf("\nTheoretical Var: %.4f\n", theoretical_var))
cat(sprintf("Simulated Var  : %.4f\n", variance_est))
cat(sprintf("Ratio          : %.4f (should be ~1.0)\n",
            variance_est / theoretical_var))

# Assertion: bias should be negligible
stopifnot(abs(bias) < 0.5 * sqrt(theoretical_var))
cat("\n[PASS] Unbiasedness verified.\n")

# ============================================================
# Production: survey package comparison
# ============================================================

library(survey)

set.seed(42)
idx <- sample(1:N, n, replace = FALSE)
sample_data <- population[idx, ]

design <- svydesign(
  id      = ~1,
  fpc     = rep(N, n),
  data    = sample_data
)

est <- svymean(~income, design)
manual_mean <- mean(sample_data$income)

cat(sprintf("\nManual mean : %.4f\n", manual_mean))
cat(sprintf("svymean     : %.4f\n", coef(est)))

stopifnot(abs(coef(est) - manual_mean) < 1e-10)
cat("[PASS] Manual and survey package estimates match.\n")
